File: TCR24CS042 - Kasinathan A B/web/src/data/test-cases.js
/**
 * Test case data — mirrors the 6 test cases from tests/run_tests.cpp plus an
 * interactive sandbox environment for live bad-state toggling and rerouting.
 */

export const testCases = [
  {
    id: 0,
    name: "Sandbox: Branching Paths",
    description: "Dual independent routes (S→A1→A2→G vs S→B1→B2→G) for live bad-state rerouting",
    proves: "Live bad-state toggling and dynamic path rerouting",
    isSandbox: true,
    problem: {
      initialState: 1,
      goalState: 6,
      badStates: [],
      states: [
        { id: 1, embedding: [0, 1] },   // Start (S)
        { id: 2, embedding: [1, 2] },   // Route A - State A1
        { id: 3, embedding: [2, 2] },   // Route A - State A2
        { id: 4, embedding: [1, 0] },   // Route B - State B1
        { id: 5, embedding: [2, 0] },   // Route B - State B2
        { id: 6, embedding: [3, 1] },   // Goal (G)
      ],
      transitions: [
        // Route A (Top) — total cost 3.0
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 3, to: 6, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        // Route B (Bottom) — total cost 3.6
        { id: 4, from: 1, to: 4, cost: 1.2, safety: 1.0, reliability: 1.0, available: true },
        { id: 5, from: 4, to: 5, cost: 1.2, safety: 1.0, reliability: 1.0, available: true },
        { id: 6, from: 5, to: 6, cost: 1.2, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: null,
  },
  {
    id: 1,
    name: "Basic Reachability",
    description: "Finds the unique valid path S→A→B→G",
    proves: "Finds the unique valid path",
    problem: {
      initialState: 1,
      goalState: 4,
      badStates: [],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0] },
        { id: 3, embedding: [2, 0] },
        { id: 4, embedding: [3, 0] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 3, to: 4, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: null,
  },
  {
    id: 2,
    name: "Bad State Avoidance",
    description: "Never routes through a bad state, even if that path exists",
    proves: "Never routes through a bad state",
    problem: {
      initialState: 1,
      goalState: 5,
      badStates: [3],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0] },
        { id: 3, embedding: [2, 0] },
        { id: 4, embedding: [0, 1] },
        { id: 5, embedding: [2, 1] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 3, to: 5, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 4, from: 1, to: 4, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 5, from: 4, to: 5, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: null,
  },
  {
    id: 3,
    name: "Safety Margin Trade-off",
    description: "Picks a costlier path when it stays farther from danger",
    proves: "Balances cost against distance from bad states",
    problem: {
      initialState: 1,
      goalState: 4,
      badStates: [5],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0.1] },
        { id: 3, embedding: [1, 3] },
        { id: 4, embedding: [2, 0] },
        { id: 5, embedding: [1, 0] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 4, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 1, to: 3, cost: 3.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 4, from: 3, to: 4, cost: 3.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: null,
  },
  {
    id: 4,
    name: "Dynamic Transition Removal",
    description: "Replans around a disabled edge (A→G becomes unavailable)",
    proves: "Replans around a disabled edge",
    problem: {
      initialState: 1,
      goalState: 3,
      badStates: [],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0] },
        { id: 3, embedding: [2, 0] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 1, to: 3, cost: 5.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: {
      type: "disableTransition",
      label: "Disable edge A→G (transition 2)",
      transitionId: 2,
      fromState: 2,
      toState: 3,
    },
  },
  {
    id: 5,
    name: "Goal Update",
    description: "Replans for a new goal without a full rebuild",
    proves: "Efficient incremental replanning on goal change",
    problem: {
      initialState: 1,
      goalState: 3,
      badStates: [],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0] },
        { id: 3, embedding: [2, 0] },
        { id: 4, embedding: [3, 0] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 3, to: 4, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: {
      type: "changeGoal",
      label: "Change goal to state 4",
      newGoal: 4,
    },
  },
  {
    id: 6,
    name: "Transition Addition",
    description: "Discovers a new shortcut once available (direct S→G)",
    proves: "Finds a shorter path after edge addition",
    problem: {
      initialState: 1,
      goalState: 4,
      badStates: [],
      states: [
        { id: 1, embedding: [0, 0] },
        { id: 2, embedding: [1, 0] },
        { id: 3, embedding: [2, 0] },
        { id: 4, embedding: [3, 0] },
      ],
      transitions: [
        { id: 1, from: 1, to: 2, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 2, from: 2, to: 3, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
        { id: 3, from: 3, to: 4, cost: 1.0, safety: 1.0, reliability: 1.0, available: true },
      ],
    },
    replanAction: {
      type: "addTransition",
      label: "Add shortcut S→G (cost 1.5)",
      transition: {
        id: 4, from: 1, to: 4, cost: 1.5, safety: 1.0, reliability: 1.0, available: true,
      },
    },
  },
];

File: TCR24CS013 - Anna Tomy/test_cases.py
"""
test_cases.py
The 6 illustrative test cases from the assignment, written as runnable functions.
Each function builds its own small PlanningProblem, runs the DStarLite planner,
checks the result against what the assignment says SHOULD happen, and prints a
pass/fail line.
"""

from models import State, Transition, PlanningProblem
from dstar_lite import DStarLite


def _check(label: str, condition: bool):
    status = "PASS" if condition else "FAIL"
    print(f"  [{status}] {label}")
    return condition


# ---------------------------------------------------------------------------
# Test Case 1: Basic Reachability   S -> A -> B -> G
# ---------------------------------------------------------------------------
def test_case_1():
    print("\nTest Case 1: Basic Reachability (S -> A -> B -> G)")

    states = [State(0, [0, 0]), State(1, [1, 0]), State(2, [2, 0]), State(3, [3, 0])]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=1.0, reliability=1.0),
        Transition(1, 1, 2, cost=1, safety=1.0, reliability=1.0),
        Transition(2, 2, 3, cost=1, safety=1.0, reliability=1.0),
    ]
    problem = PlanningProblem(initial_state=0, goal_state=3, bad_states=[],
                               states=states, transitions=transitions)

    planner = DStarLite()
    result = planner.plan(problem)

    print(f"  path: {result.state_path}, cost: {result.total_cost}")
    ok = _check("planner succeeds", result.success)
    ok &= _check("path is the unique route [0,1,2,3]", result.state_path == [0, 1, 2, 3])
    return ok


# ---------------------------------------------------------------------------
# Test Case 2: Bad State Avoidance
#   S -> A -> X(bad) -> G   (shorter, forbidden)
#   S -> C -> D -> G        (longer, must be chosen)
# ---------------------------------------------------------------------------
def test_case_2():
    print("\nTest Case 2: Bad State Avoidance")

    states = [
        State(0, [0, 0]),   # S
        State(1, [1, 1]),   # A
        State(2, [2, 1]),   # X (bad)
        State(3, [1, -1]),  # C
        State(4, [2, -1]),  # D
        State(5, [3, 0]),   # G
    ]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=1.0, reliability=1.0),  # S->A
        Transition(1, 1, 2, cost=1, safety=1.0, reliability=1.0),  # A->X
        Transition(2, 2, 5, cost=1, safety=1.0, reliability=1.0),  # X->G
        Transition(3, 0, 3, cost=1, safety=1.0, reliability=1.0),  # S->C
        Transition(4, 3, 4, cost=1, safety=1.0, reliability=1.0),  # C->D
        Transition(5, 4, 5, cost=1, safety=1.0, reliability=1.0),  # D->G
    ]
    problem = PlanningProblem(initial_state=0, goal_state=5, bad_states=[2],
                               states=states, transitions=transitions)

    planner = DStarLite()
    result = planner.plan(problem)

    print(f"  path: {result.state_path}, cost: {result.total_cost}")
    ok = _check("planner succeeds", result.success)
    ok &= _check("bad state (2) never visited", 2 not in result.state_path)
    ok &= _check("path is the safe route [0,3,4,5]", result.state_path == [0, 3, 4, 5])
    return ok


# ---------------------------------------------------------------------------
# Test Case 3: Safety Margin
#   Path 1 (via A): cheap (raw cost 2), but its edges have LOW safety scores
#                   because it passes right next to a bad state.
#   Path 2 (via C): pricier (raw cost 4), but its edges have HIGH safety scores
#                   because it stays far from the bad state.
#   We show that safety_weight controls which one the planner actually picks.
# ---------------------------------------------------------------------------
def test_case_3():
    print("\nTest Case 3: Safety Margin (cost vs. danger trade-off)")

    states = [
        State(0, [0, 0]),      # S
        State(1, [1, 0.05]),   # A -- sits right next to the bad state
        State(2, [2, 0]),      # G
        State(3, [1, 5]),      # C -- far from the bad state
        State(4, [1, 0]),      # BAD state
    ]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=0.1, reliability=1.0),  # S->A  (cheap, risky)
        Transition(1, 1, 2, cost=1, safety=0.1, reliability=1.0),  # A->G  (cheap, risky)
        Transition(2, 0, 3, cost=2, safety=0.9, reliability=1.0),  # S->C  (pricier, safe)
        Transition(3, 3, 2, cost=2, safety=0.9, reliability=1.0),  # C->G  (pricier, safe)
    ]
    problem = PlanningProblem(initial_state=0, goal_state=2, bad_states=[4],
                               states=states, transitions=transitions)

    # Pure cost minimization: safety_weight = 0 -> should take the cheap, risky route
    cheap_planner = DStarLite(safety_weight=0.0, reliability_weight=0.0)
    cheap_result = cheap_planner.plan(problem)
    print(f"  safety_weight=0.0 -> path={cheap_result.state_path}, "
          f"cost={cheap_result.total_cost}, min-dist-to-bad={cheap_result.safety_score:.2f}")

    # Safety-sensitive: safety_weight high -> should switch to the safer, pricier route
    safe_planner = DStarLite(safety_weight=0.5, reliability_weight=0.0)
    safe_result = safe_planner.plan(problem)
    print(f"  safety_weight=0.5 -> path={safe_result.state_path}, "
          f"cost={safe_result.total_cost}, min-dist-to-bad={safe_result.safety_score:.2f}")

    ok = _check("both planners succeed", cheap_result.success and safe_result.success)
    ok &= _check("safety_weight=0 picks the cheap route via A",
                 cheap_result.state_path == [0, 1, 2])
    ok &= _check("safety_weight=0.5 switches to the safer route via C",
                 safe_result.state_path == [0, 3, 2])
    ok &= _check("the safer route does keep a larger minimum distance from the bad state",
                 safe_result.safety_score > cheap_result.safety_score)
    return ok


# ---------------------------------------------------------------------------
# Test Case 4: Dynamic Transition
#   Initially S -> A -> G is the plan. Then (A, G) becomes unavailable.
#   Planner must find the alternative route WITHOUT a full graph rebuild.
# ---------------------------------------------------------------------------
def test_case_4():
    print("\nTest Case 4: Dynamic Transition (edge disappears mid-plan)")

    states = [State(0, [0, 0]), State(1, [1, 0]), State(2, [2, 0]),
              State(3, [1, -1]), State(4, [2, -1])]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=1.0, reliability=1.0),  # S->A
        Transition(1, 1, 2, cost=1, safety=1.0, reliability=1.0),  # A->G  (will break)
        Transition(2, 0, 3, cost=1, safety=1.0, reliability=1.0),  # S->C
        Transition(3, 3, 4, cost=1, safety=1.0, reliability=1.0),  # C->D
        Transition(4, 4, 2, cost=1, safety=1.0, reliability=1.0),  # D->G
    ]
    problem = PlanningProblem(initial_state=0, goal_state=2, bad_states=[],
                               states=states, transitions=transitions)

    planner = DStarLite()
    result1 = planner.plan(problem)
    print(f"  before break: path={result1.state_path}, cost={result1.total_cost}")

    result2 = planner.update_transition(1, 2, available=False)
    print(f"  after  break: path={result2.state_path}, cost={result2.total_cost}, "
          f"states touched this replan={result2.explored_states - result1.explored_states}")

    ok = _check("initial plan succeeds via A", result1.success and result1.state_path == [0, 1, 2])
    ok &= _check("replan succeeds via alternative route", result2.success and 1 not in result2.state_path[1:])
    ok &= _check("replan touches far fewer states than a full re-search would",
                 (result2.explored_states - result1.explored_states) < len(states))
    return ok


# ---------------------------------------------------------------------------
# Test Case 5: Goal Update
# ---------------------------------------------------------------------------
def test_case_5():
    print("\nTest Case 5: Goal Update")

    states = [State(0, [0, 0]), State(1, [1, 0]), State(2, [2, 0]), State(3, [3, 0])]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=1.0, reliability=1.0),
        Transition(1, 1, 2, cost=1, safety=1.0, reliability=1.0),
        Transition(2, 2, 3, cost=1, safety=1.0, reliability=1.0),
    ]
    problem = PlanningProblem(initial_state=0, goal_state=2, bad_states=[],
                               states=states, transitions=transitions)

    planner = DStarLite()
    result1 = planner.plan(problem)
    print(f"  goal=2: path={result1.state_path}")

    result2 = planner.set_goal(3, problem)
    print(f"  goal=3: path={result2.state_path}")

    ok = _check("first goal reached correctly", result1.state_path == [0, 1, 2])
    ok &= _check("revised goal reached correctly", result2.state_path == [0, 1, 2, 3])
    return ok


# ---------------------------------------------------------------------------
# Test Case 6: Transition Addition (a new shortcut appears)
# ---------------------------------------------------------------------------
def test_case_6():
    print("\nTest Case 6: Transition Addition (shortcut appears)")

    states = [State(0, [0, 0]), State(1, [1, 0]), State(2, [2, 0]), State(3, [3, 0])]
    transitions = [
        Transition(0, 0, 1, cost=1, safety=1.0, reliability=1.0),
        Transition(1, 1, 2, cost=1, safety=1.0, reliability=1.0),
        Transition(2, 2, 3, cost=1, safety=1.0, reliability=1.0),
    ]
    problem = PlanningProblem(initial_state=0, goal_state=3, bad_states=[],
                               states=states, transitions=transitions)

    planner = DStarLite()
    result1 = planner.plan(problem)
    print(f"  before shortcut: path={result1.state_path}, cost={result1.total_cost}")

    shortcut = Transition(99, 0, 3, cost=1, safety=1.0, reliability=1.0)  # S -> G directly
    result2 = planner.add_transition(shortcut)
    print(f"  after  shortcut: path={result2.state_path}, cost={result2.total_cost}")

    ok = _check("original plan takes 3 hops", len(result1.state_path) == 4)
    ok &= _check("planner discovers and uses the new shortcut", result2.state_path == [0, 3])
    ok &= _check("shortcut reduces total cost", result2.total_cost < result1.total_cost)
    return ok


if __name__ == "__main__":
    tests = [test_case_1, test_case_2, test_case_3, test_case_4, test_case_5, test_case_6]
    results = [t() for t in tests]

    print("\n" + "=" * 40)
    print(f"SUMMARY: {sum(results)}/{len(results)} test cases passed")
    print("=" * 40)